tests/integration-tests/src/test/java/org/apache/activemq/artemis/tests/integration/server/RetroactiveAddressTest.java

```diff
@@ -0,0 +1,287 @@
+      producer.send(message);
+      producer.close();
+      Wait.assertTrue(() -> server.locateQueue(divertQueue).getMessageCount() == 1, 500, 50);
+
+      server.stop();
+      server.start();
+      assertNotNull(server.locateQueue(divertQueue));
+      Wait.assertTrue(() -> server.locateQueue(divertQueue).getMessageCount() == 1, 500, 50);
+      server.getAddressSettingsRepository().addMatch(addressName.toString(), new AddressSettings().setRetroactiveMessageCount(10));
+      locator = createInVMNonHALocator();
+      sf = createSessionFactory(locator);
+      session = addClientSession(sf.createSession(false, true, true));
+
+      producer = session.createProducer(addressName);
+      message = session.createMessage(true);
+      message.getBodyBuffer().writeString(data + "2");
+      producer.send(message);
+      producer.close();
+
+      session.createQueue(addressName.toString(), RoutingType.ANYCAST, queueName1.toString());
+      Wait.assertTrue(() -> server.locateQueue(queueName1) != null);
+      Wait.assertTrue(() -> server.locateQueue(queueName1).getMessageCount() == 2, 500, 50);
+
+      ClientConsumer consumer = session.createConsumer(queueName1);
+      session.start();
+      message = consumer.receive(1000);
+      assertNotNull(message);
+      message.acknowledge();
+      assertEquals(data + "1", message.getBodyBuffer().readString());
+      message = consumer.receive(1000);
+      assertNotNull(message);
+      message.acknowledge();
+      assertEquals(data + "2", message.getBodyBuffer().readString());
+      consumer.close();
+      Wait.assertTrue(() -> server.locateQueue(queueName1).getMessageCount() == 0, 500, 50);
+
+      Wait.assertTrue(() -> server.locateQueue(divertQueue).getMessageCount() == 2, 2000, 100);
+   }
+
+   @Test
+   public void testUpdateAfterRestart() throws Exception {
+      final int COUNT = 10;
+      final SimpleString addressName = SimpleString.toSimpleString("myAddress");
+      final SimpleString divertQueue = ResourceNames.getRetroactiveResourceName(server.getConfiguration().getInternalNamingPrefix(), addressName, ResourceNames.QUEUE);
+      server.getAddressSettingsRepository().addMatch(addressName.toString(), new AddressSettings().setRetroactiveMessageCount(COUNT));
+      server.addAddressInfo(new AddressInfo(addressName));
+      Wait.assertTrue(() -> server.locateQueue(divertQueue).getRingSize() == COUNT, 1000, 100);
+      server.stop();
+      server.start();
+      server.getAddressSettingsRepository().addMatch(addressName.toString(), new AddressSettings().setRetroactiveMessageCount(COUNT * 2));
+      Wait.assertTrue(() -> server.locateQueue(divertQueue).getRingSize() == COUNT * 2, 1000, 100);
+      server.getAddressSettingsRepository().addMatch(addressName.toString(), new AddressSettings().setRetroactiveMessageCount(COUNT));
+      Wait.assertTrue(() -> server.locateQueue(divertQueue).getRingSize() == COUNT, 1000, 100);
+   }
+
+   @Test
```

Comment:
Should add a test that an anycast producer produces messages and ensures they dont get retroactively consumed by on multicast